Automattic/woocommerce-payments
# Fix WooPay Multiple Redirects

#### Changes proposed in this Pull Request




This is one piece of the fix for https://github.com/Automattic/woopay/issues/2125. This fix prevents cases where WooPay sending multiple redirect messages will trigger too many WooPay session init calls and create too many draft orders causing issues with WooPay checkout.

This change was needed because there was a scenario where 3rd party cookies were not blocked and the shopper was already logged into WooPay where the `redirect_to_woopay` post message was getting posted to the merchant store multiple times. This triggered multiple session init calls which created an issue with draft orders being created on the merchant store and at times preventing WooPay checkout.

This PR addresses that issue by checking if we are already in the process of redirecting the shopper and prevents any further redirects. It does this check by using a flag within the WooPay front end specific code.

Below are the network logs from dev tools after clicking the WooPay button while already logged into WooPay.

**Before**
![too-many-redirects](https://github.com/Automattic/woocommerce-payments/assets/68524302/e207e029-6c2c-4158-bdb2-d592bff7bfcb)

**After**
![Screenshot 2023-09-15 at 10 28 35 AM](https://github.com/Automattic/woocommerce-payments/assets/68524302/5984deca-0e78-4731-884e-4977b754d9b3)





#### Testing instructions





1. Make sure 3rd party cookies are allowed (or if you're using the same TLD for both merchant and platform, it will work either way).
2. Go to the merchant store and checkout using WooPay.
3. Stay logged into WooPay and immediately go to checkout again.
4. This time when the OTP modal opens it should begin an immediate redirect.
5. You should only see 1 wcpay_init_woopay call.
6. At this point check in the merchant store admin area and confirm this is only one draft order created.
7. Continue checkout on WooPay successfully.
8. Confirm that the only draft order from before is now completed and there were no additional draft orders created for that session.

-------------------


**Post merge**
